impl/sql/mutationstorage/queue_test.go

```diff
@@ -0,0 +1,92 @@
+// Copyright 2018 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package mutationstorage
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	_ "github.com/mattn/go-sqlite3"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	pb "github.com/google/keytransparency/core/api/v1/keytransparency_go_proto"
+)
+
+func TestSend(t *testing.T) {
+	ctx := context.Background()
+	db := newDB(t)
+	m, err := New(db)
+	if err != nil {
+		t.Fatalf("Failed to create mutations: %v", err)
+	}
+	domainID := "foo"
+	update := []byte("bar")
+	ts1 := time.Now()
+	ts2 := ts1.Add(time.Duration(1))
+	ts3 := ts2.Add(time.Duration(1))
+
+	for _, tc := range []struct {
+		desc     string
+		ts       time.Time
+		wantCode codes.Code
+	}{
+		// Enforce timestamp uniqueness.
+		{desc: "First", ts: ts2},
+		{desc: "Second", ts: ts2, wantCode: codes.Aborted},
+		// Enforce a monotonically increasing timestamp
+		{desc: "Old", ts: ts1, wantCode: codes.Aborted},
+		{desc: "New", ts: ts3},
+	} {
+		err := m.send(ctx, domainID, update, tc.ts)
+		if got, want := status.Code(err), tc.wantCode; got != want {
+			t.Errorf("%v: send(): %v, got: %v, want %v", tc.desc, err, got, want)
+		}
+	}
+}
+
+func TestWatermark(t *testing.T) {
+	ctx := context.Background()
+	db := newDB(t)
+	m, err := New(db)
+	if err != nil {
+		t.Fatalf("Failed to create mutations: %v", err)
+	}
+	domainID := "foo"
+
+	for _, tc := range []struct {
+		desc      string
+		send      bool
+		wantAfter int64
+	}{
+		{desc: "no rows", wantAfter: 0},
+		{desc: "first", send: true, wantAfter: time.Now().UnixNano()},
+		{desc: "second", send: true, wantAfter: time.Now().UnixNano()},
+	} {
```

Comment:
The test case structure doesn't add anything for this particular test -- a better test would be to just have a loop up to N doing {HighWatermark, check wantAfter, update wantAfter, Send}.

Reply:
The test case structure makes this a bit easier to read and modify later.

Reply:
Yes, but using this structure means that the `wantAfter` test is pretty useless for the final case -- it's going to have (nearly) the same value as the test before, and so you're not checking that the second addition happens after the first, you're checking that the second addition happens after test-start.

Reply:
`wantAfter` is indeed pretty useless.  
I've reworked the tests to set and expect exact watermarks. Does that help?

Reply:
Yep, splendid.